pkg/services/ngalert/store/instance_database.go

```diff
@@ -88,43 +89,89 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
 	})
 }
 
-// SaveAlertInstance is a handler for saving a new alert instance.
-func (st DBstore) SaveAlertInstance(ctx context.Context, cmd *models.SaveAlertInstanceCommand) error {
-	return st.SQLStore.WithDbSession(ctx, func(sess *sqlstore.DBSession) error {
-		labelTupleJSON, labelsHash, err := cmd.Labels.StringAndHash()
+// SaveAlertInstances saves all the provided alert instances to the store in a single transaction.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd ...models.AlertInstance) error {
+	// The function starts a single transaction and batches writes into
+	// statements with `maxRows` instances per statements. This makes for a
+	// fairly efficient transcation without creating statements that are too long
+	// for some databases to process. For example, SQLite has a limit of 999
+	// variables per write.
+
+	err := st.SQLStore.WithTransactionalDbSession(ctx, func(sess *sqlstore.DBSession) error {
+		keyNames := []string{"rule_org_id", "rule_uid", "labels_hash"}
+		fieldNames := []string{
+			"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state",
+			"current_reason", "current_state_since", "current_state_end", "last_eval_time",
+		}
+		fieldsPerRow := len(fieldNames)
+		maxRows := 20
+		maxArgs := maxRows * fieldsPerRow
+
+		// Prepare a statement for the maximum batch size.
+		bigUpsertSQL, err := st.SQLStore.Dialect.UpsertMultipleSQL(
+			"alert_instance", keyNames, fieldNames, maxRows)
 		if err != nil {
 			return err
 		}
 
-		alertInstance := &models.AlertInstance{
-			RuleOrgID:         cmd.RuleOrgID,
-			RuleUID:           cmd.RuleUID,
-			Labels:            cmd.Labels,
-			LabelsHash:        labelsHash,
-			CurrentState:      cmd.State,
-			CurrentReason:     cmd.StateReason,
-			CurrentStateSince: cmd.CurrentStateSince,
-			CurrentStateEnd:   cmd.CurrentStateEnd,
-			LastEvalTime:      cmd.LastEvalTime,
+		bigStmt, err := sess.DB().Prepare(bigUpsertSQL)
```

Comment:
I think this is missing a `close()`, same below on 156th `stmt, err := sess.DB().Prepare(upsertSQL)`. The session closure isn't cascading to  the statements because they are prepared via direct DB access rather than via `xorm.Session.Prepare()` @joeblubaugh